src/JustSaying/Messaging/Channels/Receive/MessageReceiveBuffer.cs

```diff
@@ -0,0 +1,164 @@
+        private string _backoffStrategyName;
+
+        public ChannelReader<IQueueMessageContext> Reader => _channel.Reader;
+
+        public MessageReceiveBuffer(
+            int prefetch,
+            int bufferSize,
+            TimeSpan readTimeout,
+            ISqsQueue sqsQueue,
+            MiddlewareBase<GetMessagesContext, IList<Message>> sqsMiddleware,
+            IMessageMonitor monitor,
+            ILogger<IMessageReceiveBuffer> logger,
+            IMessageBackoffStrategy messageBackoffStrategy = null)
+        {
+            _channel = Channel.CreateBounded<IQueueMessageContext>(bufferSize);
+            _prefetch = prefetch;
+            _bufferSize = bufferSize;
+            _readTimeout = readTimeout;
+            _sqsQueue = sqsQueue ?? throw new ArgumentNullException(nameof(sqsQueue));
+            _sqsMiddleware = sqsMiddleware ?? throw new ArgumentNullException(nameof(sqsMiddleware));
+            _monitor = monitor ?? throw new ArgumentNullException(nameof(monitor));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _backoffStrategyName = messageBackoffStrategy?.GetType()?.Name;
+
+            if (messageBackoffStrategy != null)
+            {
+                _requestMessageAttributeNames.Add(MessageSystemAttributeName.ApproximateReceiveCount);
+            }
+        }
+
+        /// <summary>
+        /// Starts the receive buffer until it's cancelled by the stopping token.
+        /// </summary>
+        /// <param name="stoppingToken">A <see cref="CancellationToken"/> that will stop the buffer when signalled.</param>
+        /// <returns>A <see cref="Task{TResult}"/> representing the asynchronous operation to receive the messages.</returns>
+        public async Task Run(CancellationToken stoppingToken)
+        {
+            await Task.Yield();
+
+            ChannelWriter<IQueueMessageContext> writer = _channel.Writer;
+            try
+            {
+                while (true)
+                {
+                    stoppingToken.ThrowIfCancellationRequested();
+
+                    using (_monitor.MeasureThrottle())
+                    {
+                        bool canWrite = await writer.WaitToWriteAsync(stoppingToken).ConfigureAwait(false);
+                        if (!canWrite) break;
+                    }
+
+                    IList<Message> messages;
+                    using (_monitor.MeasureReceive(_sqsQueue.QueueName, _sqsQueue.RegionSystemName))
+                    {
+                        messages = await GetMessagesAsync(_bufferSize, stoppingToken).ConfigureAwait(false);
+
+                        if (messages == null) continue;
+                    }
+
+                    foreach (Message message in messages)
+                    {
+                        IQueueMessageContext messageContext = _sqsQueue.ToMessageContext(message);
+                        await writer.WriteAsync(messageContext).ConfigureAwait(false);
```

Comment:
Probably want to pass `stoppingToken` in here

Reply:
Ah thanks, good catch. Fixed 👍